src/librustdoc/markdown.rs

```diff
@@ -89,12 +98,44 @@ pub fn render(input: &str, mut output: PathBuf, matches: &getopts::Matches,
 
     reset_ids(false);
 
-    let rendered = if include_toc {
-        format!("{}", MarkdownWithToc(text, render_type))
+    let (hoedown_output, pulldown_output) = if include_toc {
+        // Save the state of USED_ID_MAP so it only gets updated once even
+        // though we're rendering twice.
+        let orig_used_id_map = USED_ID_MAP.with(|map| map.borrow().clone());
+        let hoedown_output = format!("{}", MarkdownWithToc(text, RenderType::Hoedown));
+        USED_ID_MAP.with(|map| *map.borrow_mut() = orig_used_id_map);
+        let pulldown_output = format!("{}", MarkdownWithToc(text, RenderType::Pulldown));
```

Comment:
Is there a way we can somehow combine this chunk of code with the one below and with https://github.com/GuillaumeGomez/rust/blob/92c38d31e4f3797517d1a33619eb2628758cf464/src/librustdoc/html/render.rs#L1838-L1843 so we don't have three copies of the "render this markdown twice" code? (I'm thinking some kind of function that takes a `RenderType` and returns a string so it can also be used right here?)

Reply:
I was thinking about it as well but every solution seems more like a boilerplate than a real solution. I'll give it a deeper look.

Reply:
So my thinking is something like this, in `html/render.rs` so it can be used alongside `USED_ID_MAP`:

```rust
fn render_text<F: FnMut(RenderType) -> String>(render: F) -> (String, String) {
    // (original comment about saving USED_ID_MAP)
    let orig_used_id_map = USED_ID_MAP.with(|map| map.borrow().clone());
    let hoedown_output = render(text, RenderType::Hoedown);
    USED_ID_MAP.with(|map| *map.borrow_mut() = orig_used_id_map);
    let pulldown_output = render(text, RenderType::Pulldown);
    (hoedown_output, pulldown_output)
}
```

...and then it could be called like this:

```rust
render_text(|type| format!("{}", Markdown(text, type)));
render_text(|type| format!("{}", MarkdownWithToc(text, type)));
```

Reply:
Totally seems like a boilerplate haha. I'll give it a try. ;)

Reply:
I don't see how refactoring this to combine the code would be called "boilerplate", but if you think it would be too much effort i can just drop it.

Reply:
Can you export `render_text` and use it here? (That's why i brought it up in the first place. `>_>`)